### DISCUSSION TITLE
Alignment based on probing

### DISCUSSION THREAD

**Discussion by Klippert** [terjeio/grblHAL#288] [Q&A]
I know this might be a very long shot. But - i was playing around with fusion part alignment, and watched this video: https://www.youtube.com/watch?v=y2D1Ct8nMEM
And that is just the most brilliant thing i've seen in a very long time! It would so much take my headaches away when trying to get a part back on the table to trim a bit off here, make some threaded holes there, or engrave something there. (or possibly even make a threaded hole deeper 8-))

So my big question is, does any of you have any experience with this kind of usage?
And is it even remotely likely to add this to grbl-hal?

**Comment by phil-barrett** [comment]
Makes me want a 5 Axis machine!  Given my already long ADD inspired project list, it will not be any time soon!

I think that kind of engraving can be done with grblHAL/ioSender in a 3 Axis machine if the part was correctly aligned along the Y Axis. The key is a holder that aligns the part properly. Set the probe region to the exact area to be engraved and then probe for the heightmap using a fairly fine grid (1mm?).  Apply the map to the engraving GCode and it should work.  It probably will require some experimentation.

**Reply by terjeio** [reply]
> And is it even remotely likely to add this to grbl-hal?

It believe is not going to be added to grblHAL as it, IMO, has to be done in the sender.

> I think that kind of engraving can be done with grblHAL/ioSender in a 3 Axis machine if the part was correctly aligned along the Y Axis.

I think so too.

**Reply by Klippert** [reply]
Ok. so all that need to be done in grbl-hal is that it has to support gcode used for probing (from fusion) - and i'm not sure if that's not already fully supported. - I guess there might also be a gap with the GRBL post processor in fusion...
It would be rather cool, if it could be made to work.

**Reply by terjeio** [reply]
> Ok. so all that need to be done in grbl-hal is that it has to support gcode used for probing (from fusion) - and i'm not sure if that's not already fully supported.

What has to be done is to enable the sender to excecute the probing gcode, catch the probing data and export it in a format that Fusion 360 expects. If I understand correctly Fusion360 then transforms the original gcode.

But then from the video comments:

"Correct me if I'm wrong but isn't this a Fusion 360 option that costs an additional $1500 per year?"

The answer was yes - a tad expensive for grblHAL users?

Anyway this is a bit like heightmap probing and gcode transform (for the Z-axis only) done in ioSender. If there are anyone out there able to implement a similar transform algorithm that Fusion360 does I believe I can integrate that in the sender. The sender tokenizes the gcode on load so that part is taken care of already...

**Reply by Klippert** [reply]
that's true. i'm using an educational license, so it's all included for free. But not everyone might be that lucky. 
In any case, there shouldn't be anything stopping one from posting probing commands to grbl-gcode, right? 
Or is probing using special M and G codes?

regarding building the algorithm into iosender, that would be amazing, however i guess when you're talking rotational alignment it gets a lot more complex than a heightmap? (we need someone with the whacking great math hat here)...
